int16 p0_x
int16 p0_y
int16 p1_x
int16 p1_y
int16 p2_x
int16 p2_y
int16 p3_x
int16 p3_y